### ISSUE TITLE
Improve either PyWorker documentation or internals in general

### ISSUE THREAD

**Issue by WebReflection** [pyscript/pyscript#1980]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

This issue https://github.com/pyscript/pyscript/issues/1979 mentioned a few things that are kinda correct:

  * our documentation around *PyWorker* does not cover the whole story
  * our *JS* implementation differs from the *Python* related one

Please allow me to elaborate on latter point!

### Different Implementation

After recent real-world use cases on how our *PyWorker* and recently the *MPWorker*, land on the main thread, out of utilities offered by the module, it's clear we need to be aligned around the `from pyscript import PyWorker` story:

  * we have two different `async` utilities on the JS side, which solve all the things but are not reflected in *Python* side
  * we have two different names that might be handy disambiguate on *Python* side too
  * the *JS* exported utility does more because it brings in config defaults and it grants `sync` related utilities can be safely attached and used elsewhere
  * the *Python* utility is rather a waw `XWorker` wrap from *polyscript*, making the whole bootstrap awkward when we export also proper utilities on main ...
  * the async nature of *PyWorker* and *MPWorker* on main are kinda dictated by the fact we can't block core with top level await because plugins use core itself to bootstrap, and that would create a bundling cross-reference shnanigan with repeated, and bloated, unnecessary code all over ...

### Proposals

In order to provide the best UX/DX we can, I am considering these approaches:

  * we remove *PyWorker* all together from the module, as breaking change, and we suggest that if any advanced usage of the underlying `XWorker` is needed, users can just target that script and access its `xworker` field instead, as explained and tested to work in the original issue
  * we change both *PyWorker* and *MPWorker* utilities from core, making these **async** classes by default, following what we shipped on the *JS* side (imho, ugly)
  * we orchestrate everything via a *Python* proxy that actually executes all pending operations only once the worker is ready to go (imho, very complicated and error prone) ...
  * something else I haven't thought about (please share!)

This might be seen rather as a discussion, but the current API has a hole we need to either close or fill with content, so whatever we decide should likely be handled as issue already.

**Comment by JeffersGlass** [comment]
Just chiming in weeks later to say I stumbled across this confusion as well - this is, why can't I create a PyWorker from within PyScript, and what the story around that process is? I'm still not entirely sure I understand it myself, but I would like to 😀

**Comment by Neon22** [comment]
Clearly how people intend to use a Worker is probably one of the main issues because if you're firing it off from JS then its diff from inside a pyscript. Also webworkers, serviceworkers, background fetches, etc.

The usecase I am looking for is:

I have a main.py with all of my UI and interaction in it. Every now and then (based on the user's actions) I want to do a long running calc involving permutations. Could be short, could be a minute or two. So Ideally I want to `Worker` that process.
So I need to send it args and then wait for a reply. When the reply arrives, I want to redraw some stuff in my UI and update some variables.
Then do it again when the user makes another change.


What I have working so far is this:
 - must be 2024.3.2
 - worker needs the worker-options

main.py
```
from pyscript import PyWorker, display, ffi

# Workers
def draw_floats(width,height,data):
    """
    Gets called when worker returns
    """
    display(f"4 Returned with, {width}", append=True)
    print(" - got data:",width,height,len(data))
    # draw it
    print("5 calling hello again")
    worker.sync.hello = hello
    print("6 started worker hello 2")


def hello(name="world"):
    print("3 hello runs:",name)
    display(f"Hello, {name}", append=True)

if __name__ == "__main__":
    worker_options = ffi.to_js({"type": "pyodide"})
    worker = PyWorker("./worker1.py", worker_options) 
    #
    worker.sync.hello = hello
    print("1 started worker hello")
    worker.sync.draw_floats = draw_floats
    print("2 started worker draw_floats")```

worker1.py
```
from pyscript import sync

class Foo(object):
    perm_var = 0

# can use __main__ but not required
if __name__ == "__main__":
    a = 23
    foo = Foo()
    sync.draw_floats(a,1, [1,2,3,4,5,6,7,8])

    sync.hello(f"stuff sent back{foo.perm_var}")
    foo.perm_var += 1
```
its here: - 
but it might be quite dynamic and not working when you visit.

Clearly the worker can get run more than once and can send more than one sync back.
But (and its not a bad but because its probably a good but) the Worker starts from scratch when called by main.
Pretty sure its instantiating an entire pyodide to do this as it takes a while for the screen to update when its running (I'm a long way from any servers on tiny B/W)
So an mpy worker is probably the best if possible, and if my code will work in mpy.

Given there are so many Worker usecases - maybe each one could be stated and tested independently. ??

**Comment by WebReflection** [comment]
> the Worker starts from scratch when called by main.
> Pretty sure its instantiating an entire pyodide to do this as it takes a while for the screen to update when its running (I'm a long way from any servers on tiny B/W)

this is inevitable ... workers can't share the interpreter that lands on their own realm (global context, thread)

the ideal scenario to me is having mpy only on the main thread and use pyodide as worker so that any module/package can go in and that's pretty much it.

bootstrapping workers within Python and maybe from Workers already feels an overly-convoluted use case that's hard to reason about or support to me ... if you run already in a worker, why would you do/need that?

**Comment by Neon22** [comment]
Your reply doesn't seem to be related to what I wrote. Perhaps I was not clear. Sorry about that. I will restate.

I'm using pyodide in my main - because some of the things I want to do currently need Pyodide. I would use mpy in my main if I could and will do so when I can. (I can't remember right now what features I am referring to, sorry)

I want to use a worker because I have a non-urgent side task to do which will bind up main for too long. Maybe a minute - using permutations to search for a difficult solution. I am not fetching from a remote or anything like that - just running python in the worker. So no need for pyodide in my worker (unless maybe I use Sets or something else mpy does not have full support for. (thanks for getting random fixed btw))

I realise that doing basic work in a separate thread is not a traditional main use of Workers in a JS environment - perhaps better to push it back to the server - but not what I want to do.

I don't want to boot workers from within workers. Can't see the point of that but I guess it might happen if someone were to include into their 'new' pyscript app a previous solution which called a worker but they want to invoke it as a worker. and so you'd have a worker instantiating another worker. But not the usecase I'm talking about.

Its absolutely not a problem that a worker starts a new instance of pyodide . I wasn't complaining about it doing so - merely noting it. I guess I was trying to work out how you guys have implemented workers by direct experimentation as there were no examples/docs addressing my situation that I could find.
Sorry if that came across as a complaint. It isn't.

**Comment by Neon22** [comment]
BTW - @nickburns on discord solved my problem. (passing args into a worker invoked from main.py
My version of that solution is here:
 - https://pyscript.com/@neon22/simulation-worker/latest

**Comment by WebReflection** [comment]
> Sorry if that came across as a complaint. It isn't.

I am sorry if my answer was too harsh ... I want to do the right thing and so far the right thing landed in JS but not within the Python code itself ... I mean, it works, but it's not ideal and harder to explain in its dual state.

From JS you can create a PyWorker or MPWorker directly, wait for the `sync` to be usable then add/share invoke whatever utility you like. It's correct in both timing, it disambiguate through the worker name and it's been used already to bootstrap pyodide or micropython in a worker and delegate to these everything that might block the main thread instead.

> I realise that doing basic work in a separate thread is not a traditional main use of Workers in a JS environment

workers **are** traditionally used to do exactly that: offload expensive computation from the main thread and never block the UI. Specially with Pyodide taking seconds to bootstrap and blocking while doing so, workers are a perfect answer and also pyodide on the browser is safe, Python running in your server? good luck with that 😅

So it looks like we are aligned, the point now is: do we want to reflect the worker bootstrap/api in Python too? To me the answer isnot obvious but from Python I expect anyone to easily bootstrap either Pyodide or microPython workers from main and/or even from workers themselves, and this might be shaped after the current JS, breaking though what's out already (and purposely poorly documented as it's pretty experimental that part to date).

I hope this clarifies a bit more where we are and hopefully help moving forward this topic.

**Comment by JeffersGlass** [comment]
Here's the short example I've been playing with. I'm essentially copying the example from the [`sync_main_only` PR](https://github.com/pyscript/pyscript/pull/1967), but trying to do it using only Python.

```html
<script type="module" src="https://pyscript.net/releases/2024.3.2/core.js"></script>
<script type="py" async>
    from pyscript import PyWorker, ffi

    worker_options = ffi.to_js({"type": "pyodide", "config": {"sync_main_only" : True}})
    worker = await PyWorker("foo_worker.py", worker_options) 
    
    result = await worker.sync.utility(3)
    print(result) # Expect 6
</script>
```
```py
from pyscript import sync

def utility(*args):
    return 42

sync.utility = utility
```

But what I get is an error: 
```
TypeError: object pyodide.ffi.JsProxy can't be used in 'await' expression
```
I get this same error if I use mini-coi.js instad of `sync_main_only`.

----

If I change the main-thread tag to `mpy`, I get a different error:
```
Uncaught (in promise) Error: The interpreter "py" was not found. Available interpreters are: "mpy-script", "mpy", "micropython".
```

I'm not sure if this answers your question or adequately responds to you Andrea... but this is something that I think would be nice to have. That is, to spin up workers from Python (or Micropython) and use them to "do work" in a worker.

**Comment by WebReflection** [comment]
> worker = await PyWorker("foo_worker.py", worker_options)

I don't think that's how it is now in Python ... you don't want to `await` the *PyWorker*

**Comment by Neon22** [comment]
@JeffersGlass there is a working version in link above using PyWorker
Basically:
```
async def calculate(event):
   # get integers seed and iter from somewhere
   number = await worker.sync.pseudorandom(seed, iter)

worker = PyWorker("./worker1.py", worker_options)
```
with worker:
```
def pseudorandom(seed: int, iters: int):
    return seed % (2 ** 47)

sync.pseudorandom = pseudorandom
```

**Comment by WebReflection** [comment]
This will/should be tackled where it belongs: https://github.com/pyscript/docs/issues/77

Closing as docs and examples will indeed land in there.
